### PR TITLE
feat(rewards): add TOPReward reward model

### PR DESCRIPTION
**Pull request by s1lent4gnt**
## Title

feat(rewards): add TOPReward reward model

## Type / Scope

* **Type**: Feature
* **Scope**: reward models, docs, tests

## Summary / Motivation

This PR adds TOPReward as a native LeRobot reward model. TOPReward is a Qwen3-VL-based **zero-shot** video-language reward model that uses token log-probabilities from an off-the-shelf VLM as a robotic reward signal. The integration exposes TOPReward through LeRobot's standard reward-model APIs for preprocessing, inference, save/load, factory construction, tests, and documentation.

TOPReward has no fine-tuned weights of its own: the "model" is a generic VLM (default `Qwen/Qwen3-VL-8B-Instruct`) used with a fixed prompt. The LeRobot reward-model API exposes a scalar through `compute_reward()` by returning `log P("True" | video + prompt + instruction)`

## Related issues

* Related: Reward models refactor #3142

## What changed

* Added the `topreward` reward model family under `src/lerobot/rewards/topreward/` with configuration, modeling, processor, and package exports.
* Registered `reward_model.type=topreward` in the reward factory and top-level reward exports.
* Added a lightweight `TOPRewardEncoderProcessorStep` that normalises frames + task into `observation.topreward.*` entries.
* Added the inference-only `TOPRewardModel` that wraps `Qwen3VLForConditionalGeneration`, builds the chat-style prompt, and runs the VLM forward with label-masked log-prob extraction.
* Added the `topreward` optional dependency extra and lockfile updates.
* Added TOPReward documentation and reward-model card template text.
* Added focused model and processor tests covering config registration, reward extraction, HF save/load round-trips, zero-shot non-trainability, processor helper behavior, and serialization surfaces.

No breaking changes are intended. Existing reward models and policies are unchanged.

## How was this tested (or how to run locally)

* Tests added:
   * `tests/rewards/test_modeling_topreward.py`
   * `tests/rewards/test_topreward_processor.py`
* Ran the relevant tests locally:
   ```
   uv run pytest tests/rewards/test_modeling_topreward.py tests/rewards/test_topreward_processor.py -x --tb=short
   ```

Compute progress with TOPReward
```
uv run python -m lerobot.rewards.topreward.compute_rabc_weights \
    --dataset-repo-id jackvial/so101_pickplace_160_20260501a \
    --episodes 0 \
    --num-samples 15 \
    --device cuda \
    --output-path topreward_progress.parquet
```

Render the overlay
```
uv run examples/dataset/create_progress_videos.py \
    --repo-id jackvial/so101_pickplace_160_20260501a \
    --episode 0 \
    --progress-file topreward_progress.parquet
```

<p align="center">
  topreward_progress_small
</p>

## Checklist (required before merge)

- [x] Linting/formatting run (`pre-commit run -a`)
- [x] All tests pass locally (`pytest`)
- [x] Documentation updated
- [ ] CI is green

## Reviewer notes

* This PR intentionally keeps the LeRobot scope limited to TOPReward inference through the reward-model API.
* TOPReward is zero-shot, so `_save_pretrained` / `from_pretrained` / `push_model_to_hub` are overridden to skip the safetensors round-trip — a TOPReward LeRobot "checkpoint" is a single `config.json`, and the VLM weights are always identified by `vlm_name` and re-fetched from the HF Hub.
* `forward()` is intentionally inherited from `PreTrainedRewardModel`, so `is_trainable` returns `False`. This is covered by a dedicated test.
* Chat-templating and tokenisation live model-side rather than in the processor, because TOPReward's reward extraction depends on a prompt/suffix split that is only knowable post-tokenisation.

### CODE REVIEW

**Comment by jcoleharrison** [comment]
# PR Review: `feat/add-topreward-rm` → `main`

## Critical (must fix)

No findings.

## Important (strongly recommended)

### I1. `topreward` missing from the `all` optional extra

**File**: `pyproject.toml`

The `sarm` reward model is included in the `all` extra (`lerobot[sarm]`), but `topreward` is not. This means `uv sync --extra all` won't install topreward dependencies, breaking the expectation that `all` covers every in-tree feature.

**Fix**:
```toml
# In [project.optional-dependencies] -> all = [...]
# Add alongside the existing sarm entry:
"lerobot[topreward]",
```

### I2. `_pad_and_stack` has dead code — both branches of the `else` are identical

**File**: `src/lerobot/rewards/topreward/processor_topreward.py:289-291`

```python
else:
    if all(t.shape == tensors[0].shape for t in tensors):
        result[key] = torch.cat(tensors, dim=0)
    else:
        result[key] = torch.cat(tensors, dim=0)
```

Both branches do the same thing. I verified this is not a Qwen requirement — Qwen's processor doesn't need this branching. In practice `_pad_and_stack` is only called when `batch_size > 1`, which doesn't occur in the standard preprocessor pipeline (AddBatchDimensionProcessorStep produces batch=1) or in `compute_rabc_weights`. So this is dead code. If it *were* reached with different frame counts, `torch.cat` would fail on mismatched `pixel_values_videos` shapes regardless of which branch executes.

**Fix**: Collapse to a single branch with a comment:
```python
else:
    # Vision tensors (pixel_values_videos, image_grid_thw, etc.) are expected
    # to have matching shapes since max_frames is applied uniformly per batch.
    result[key] = torch.cat(tensors, dim=0)
```

### I3. Passing `labels` to VLM forward wastes compute

**File**: `src/lerobot/rewards/topreward/modeling_topreward.py:143`

```python
outputs = self.model(**inputs, labels=labels)
```

Passing `labels` causes the VLM to compute cross-entropy loss internally, but the loss is never used — only `outputs.logits` is consumed for the manual log-prob extraction. For an 8B model, this adds unnecessary computation (an extra softmax + loss reduction over the full vocabulary).

**Fix**: Don't pass labels to the model; only use them for masking locally:
```python
outputs = self.model(**inputs)
```

## Suggestions (nice to have)

### S1. Doc typo: missing newline between paper/project links

**File**: `docs/source/topreward.mdx:6-9`

The links render as a single paragraph without line breaks between them. Add blank lines or use a list:

```markdown
**Paper**: [TOPReward...](...)

**Project**: [topreward.github.io](...)

**Original code**: [github.com/TOPReward/TOPReward](...)

**Default backbone**: [Qwen/Qwen3-VL-8B-Instruct](...)
```

### S2. `_video_to_numpy` channel detection heuristic could be documented

**File**: `src/lerobot/rewards/topreward/processor_topreward.py:82-86`

The `if video.shape[1] in (1, 3)` heuristic assumes PyTorch's standard `(T, C, H, W)` layout. This is correct for LeRobot tensors but could be confusing for maintainers. A brief inline comment would help:

```python
# Assumes PyTorch-standard (T, C, H, W) layout; permutes to (T, H, W, C) for PIL.
```

## Summary

| Severity | Count | Key Themes |
|----------|-------|------------|
| Critical | 0 | — |
| Important | 3 | Missing `all` extra, dead code in padding, wasted VLM compute |
| Suggestions | 2 | Doc formatting, code clarity |

Well-structured PR that follows existing patterns (SARM). The code is clean, tests are comprehensive with proper mocking, and documentation is thorough. The important findings are straightforward to address — the most impactful are I1 (missing from `all` extra) and I3 (unnecessary loss computation on every inference call).

**Comment by s1lent4gnt** [comment]
Thank you for the suggested fixes @jcoleharrison, merged #3653

**Review by Maximellerbach — APPROVED** [review]
Looks Good !
